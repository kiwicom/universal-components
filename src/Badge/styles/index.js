// @flow

import { defaultTokens } from '@kiwicom/orbit-design-tokens';

export const wrapperColor = {
  primary: defaultTokens.paletteProductNormal,
  critical: defaultTokens.backgroundBadgeCritical,
  dark: defaultTokens.backgroundBadgeDark,
  info: defaultTokens.backgroundBadgeInfo,
  neutral: defaultTokens.backgroundBadgeNeutral,
  success: defaultTokens.backgroundBadgeSuccess,
  warning: defaultTokens.backgroundBadgeWarning,
  white: defaultTokens.backgroundBadgeWhite,
};

export const textColor = {
  primary: defaultTokens.colorTextWhite,
  critical: defaultTokens.colorTextBadgeCritical,
  dark: defaultTokens.colorTextBadgeDark,
  info: defaultTokens.colorTextBadgeInfo,
  neutral: defaultTokens.colorTextBadgeNeutral,
  success: defaultTokens.colorTextBadgeSuccess,
  warning: defaultTokens.colorTextBadgeWarning,
  white: defaultTokens.colorTextBadgeWhite,
};
