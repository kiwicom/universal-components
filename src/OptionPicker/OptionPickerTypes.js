// @flow

export type OptionType = {
  +id: string,
  +label: string,
  +text: ?string,
  +type: 'localization' | 'destination' | 'airplane' | 'bus' | 'train', // @TODO use IconNameType as enum types
  +subOptions?: Array<OptionType>,
};
