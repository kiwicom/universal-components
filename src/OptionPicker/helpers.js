// @flow

import type { OptionType } from './OptionPickerTypes';

export function getOptionsIds(options: OptionType[]) {
  return options.reduce((result, option) => {
    let ids = [option.id];
    if (option.subOptions) {
      const subOptions = option.subOptions.map(({ id }) => id);
      ids = [...ids, ...subOptions];
    }

    return [...result, ...ids];
  }, []);
}

export function areSelectedOptionsChanged(
  optionsA: OptionType[],
  optionsB: OptionType[]
) {
  const optionsIdsB = getOptionsIds(optionsB);
  const optionsIdsA = getOptionsIds(optionsA);

  return (
    optionsIdsA.length !== optionsIdsB.length ||
    !optionsIdsA.every(option => optionsIdsB.includes(option))
  );
}
