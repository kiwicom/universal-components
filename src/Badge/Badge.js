// @flow

import * as React from 'react';
import { View } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';
import Text from '../Text';
import StyleSheet from '../PlatformStyleSheet';
import type { BadgeProps, BadgeType } from './BadgeTypes';
import { textColor, wrapperColor } from './styles';

// @TODO children should be string or React element, not element in React text element
export default function Badge({
  children,
  type = 'primary',
  style,
}: BadgeProps) {
  return (
    <View style={[styles.wrapper, theme(type).wrapper, style]}>
      <Text style={[styles.text, theme(type).text]}>{children}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  wrapper: {
    alignSelf: 'flex-start',
    flexDirection: 'row',
    flexWrap: 'wrap',
    borderRadius: parseFloat(defaultTokens.borderRadiusBadge),
    paddingHorizontal: 8,
    height: parseFloat(defaultTokens.heightBadge),
    web: {
      width: 'fit-content',
    },
  },
  text: {
    fontSize: 12,
    lineHeight: 22,
    fontWeight: '500',
    letterSpacing: 0.59,
    color: defaultTokens.colorTextBadgeDark,
  },
});

const theme = (type: BadgeType = 'primary') =>
  StyleSheet.create({
    wrapper: {
      backgroundColor: wrapperColor[type],
    },
    text: {
      color: textColor[type],
    },
  });
