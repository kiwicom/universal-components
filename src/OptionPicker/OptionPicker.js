// @flow

import * as React from 'react';
import { View } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';
import type { OptionType } from './OptionPickerTypes';
import { StyleSheet } from '../PlatformStyleSheet';
import { TagsInput } from '../TagsInput';
import OptionList from './components/OptionList';
import { getOptionsIds, areSelectedOptionsChanged } from './helpers';

type Props = {|
  +label?: string,
  +placeholder?: string,
  +onChangeText: (value: string) => void,
  +options: OptionType[],
  +onPressAdd?: () => void,
  +onPressItem?: () => void,
  +onSelectedChange?: (option: Array<OptionType>) => void,
|};

type State = {|
  selected: Array<OptionType>,
  value: string,
|};

type Event = { nativeEvent: { key: string } };

const filterSubOptionsByIds = (option, ids) => {
  const { subOptions } = option;
  // check if the subOptions exists
  if (!subOptions) return option;

  // filter out selected options
  const filteredOptions = subOptions.filter(({ id }) => !ids.includes(id));
  return Object.assign({}, option, { subOptions: filteredOptions });
};

export const filterOptions = (
  options: OptionType[],
  selected: OptionType[]
) => {
  // get selectedIds
  const selectedIDs = getOptionsIds(selected);

  return options.reduce((acc = [], option) => {
    // check if the option is selected, if yes filter it out
    if (selectedIDs.includes(option.id)) {
      return acc;
    }
    return [...acc, filterSubOptionsByIds(option, selectedIDs)];
  }, []);
};

export default class OptionPicker extends React.Component<Props, State> {
  state = {
    selected: [],
    value: '',
  };

  componentDidUpdate(prevProps: Props, prevState: State) {
    const { onSelectedChange } = this.props;
    const { selected } = this.state;

    if (areSelectedOptionsChanged(selected, prevState.selected)) {
      onSelectedChange?.(selected);
    }
  }

  handlePressAdd = (option: OptionType) => {
    const { onPressAdd } = this.props;
    this.setState(
      ({ selected }) => ({ selected: [...selected, option] }),
      onPressAdd
    );
  };

  handlePressItem = (option: OptionType) => {
    const { onPressItem } = this.props;
    this.setState({ selected: [option] }, onPressItem);
  };

  handleClear = () => {
    this.setState({ value: '', selected: [] });
  };

  handleChangeText = (value: string) => {
    const { onChangeText } = this.props;
    onChangeText(value);
    this.setState({
      value,
    });
  };

  handleBackspacePress = (e: Event) => {
    const { key } = e.nativeEvent;
    const { value } = this.state;

    if (!value && key === 'Backspace') {
      this.removeLastSelectedOption();
    }
  };

  removeLastSelectedOption = () => {
    const { state } = this;

    // you can remove selected place only if at least one exists
    if (state.selected.length > 0) {
      this.setState(({ selected }) => ({
        selected: selected.slice(0, -1),
      }));
    }
  };

  render() {
    const { selected, value } = this.state;
    const { options, label, placeholder } = this.props;
    const tags = selected.map(option => option.label);

    return (
      <View>
        <View style={styles.inputWrapper}>
          <TagsInput
            autofocus
            tags={tags}
            onClearPress={this.handleClear}
            value={value}
            onChangeText={this.handleChangeText}
            onKeyPress={this.handleBackspacePress}
            label={label}
            placeholder={placeholder}
          />
        </View>
        {options && (
          <OptionList
            onItemPress={this.handlePressItem}
            onAddPress={this.handlePressAdd}
            options={filterOptions(options, selected)}
          />
        )}
      </View>
    );
  }
}

const commonStyles = {
  borderBottomWidth: 1,
  borderBottomColor: defaultTokens.borderColorInput,
};

const styles = StyleSheet.create({
  inputWrapper: {
    padding: 5,
    ios: {
      ...commonStyles,
    },
    android: {
      ...commonStyles,
    },
  },
});
