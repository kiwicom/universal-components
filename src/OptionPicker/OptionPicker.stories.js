// @flow

import * as React from 'react';
import { storiesOf } from '@storybook/react-native';
import { withKnobs, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import OptionPicker from './OptionPicker';
import mockedPlaces from './mocks/places';

storiesOf('OptionPicker', module)
  .addDecorator(getStory => getStory())
  .addDecorator(withKnobs)
  .add('Playground', () => (
    <OptionPicker
      label={text('label', 'From:') || ''}
      placeholder={text('placeholder', 'Departure point')}
      onChangeText={action('onChangeText')}
      onPressAdd={action('onPressAdd')}
      onPressItem={action('onPressItem')}
      options={mockedPlaces}
      onSelectedChange={action('change')}
    />
  ));
