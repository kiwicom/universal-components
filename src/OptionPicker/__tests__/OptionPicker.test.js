// @flow

import React from 'react';
import { shallow, render, fireEvent } from 'react-native-testing-library';
import OptionPicker, { filterOptions } from '../OptionPicker';
import options, { Prague, Berlin } from '../mocks/places';
import { Badge } from '../../Badge';

jest.mock('NativeAnimatedHelper');

test('filterOptions', () => {
  expect(filterOptions([Prague, Berlin], [Berlin])).toEqual([Prague]);
});

describe('OptionPicker', () => {
  const label = 'label';
  const placeholder = 'placeholder';
  const onChangeText = jest.fn();
  const onPressAdd = jest.fn();
  const onPressItem = jest.fn();
  const onSelectedChange = jest.fn();

  const {
    getByText,
    getByProps,
    getByTestId,
    getAllByType,
    getByPlaceholder,
  } = render(
    <OptionPicker
      label={label}
      placeholder={placeholder}
      onChangeText={onChangeText}
      options={options}
      onPressAdd={onPressAdd}
      onPressItem={onPressItem}
      onSelectedChange={onSelectedChange}
    />
  );
  test('passed the props', () => {
    expect(
      getByProps({
        onChangeText,
        options,
        onPressAdd,
        onPressItem,
        onSelectedChange,
        label,
        placeholder,
      })
    ).toBeDefined();
    expect(getByText(label)).toBeDefined();
    expect(getByPlaceholder(placeholder)).toBeDefined();
  });

  test('select place', () => {
    const Item = getByText(Prague.label);
    expect(() => getByTestId('input-tag')).toThrow('No instances found');
    fireEvent(Item, 'press');
    expect(getByTestId('input-tag')).toBeDefined();

    expect(getAllByType(Badge)).toHaveLength(1);
    expect(onPressItem).toHaveBeenCalledTimes(1);

    expect(onSelectedChange).toHaveBeenCalledTimes(1);
    expect(onSelectedChange).toHaveBeenCalledWith([Prague]);
  });

  test('select another place', () => {
    const Item = getByText(Berlin.label);
    fireEvent(Item, 'press');

    expect(onPressItem).toHaveBeenCalledTimes(2);
    expect(onSelectedChange).toHaveBeenCalledTimes(2);
    expect(onSelectedChange).toHaveBeenCalledWith([Berlin]);
    expect(getAllByType(Badge)).toHaveLength(1);
  });

  test('render', () => {
    const { output } = shallow(
      <OptionPicker
        options={options}
        onChangeText={onChangeText}
        label={label}
        placeholder={placeholder}
      />
    );

    expect(output).toMatchSnapshot();
  });
});
