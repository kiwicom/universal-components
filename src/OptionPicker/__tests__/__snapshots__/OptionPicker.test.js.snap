// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionPicker render 1`] = `
<View>
  <View
    style={
      Object {
        "borderBottomColor": "#bac7d5",
        "borderBottomWidth": 1,
        "padding": 5,
      }
    }
  >
    <TagsInput
      autofocus={true}
      fontSize={16}
      label="label"
      onChangeText={[Function]}
      onClearPress={[Function]}
      onKeyPress={[Function]}
      placeholder="placeholder"
      tags={Array []}
      value=""
    />
  </View>
  <OptionList
    onAddPress={[Function]}
    onItemPress={[Function]}
    options={
      Array [
        Object {
          "id": "prague",
          "label": "Prague",
          "subOptions": Array [
            Object {
              "id": "train-station",
              "label": "Prague Central Station",
              "text": "5 km from city center",
              "type": "train",
            },
            Object {
              "id": "airport",
              "label": "VÃ¡clav Havel Airport",
              "text": "5 km from city center",
              "type": "airplane",
            },
          ],
          "text": "Czech Republic",
          "type": "destination",
        },
        Object {
          "id": "berlin",
          "label": "Berlin",
          "subOptions": Array [
            Object {
              "id": "QPP",
              "label": "Airport of Berlin",
              "text": "4.5 km from city center",
              "type": "airplane",
            },
            Object {
              "id": "QWC",
              "label": "Zoo Railway Station",
              "text": "10 km from city center",
              "type": "airplane",
            },
          ],
          "text": "Federal Republic of Germany",
          "type": "destination",
        },
        Object {
          "id": "presov",
          "label": "Presov",
          "text": "Slovakia",
          "type": "destination",
        },
      ]
    }
  />
</View>
`;
